watch/inotify_tracker.go

```diff
@@ -166,47 +148,50 @@ func (shared *InotifyTracker) addWatch(winfo *watchInfo) error {
 		fname = filepath.Dir(fname)
 	}
 
+	var err error
 	// already in inotify watch
-	if shared.watchNums[fname] > 0 {
-		shared.watchNums[fname]++
-		if winfo.isCreate() {
-			shared.watchNums[winfo.fname]++
-		}
-		return nil
+	if shared.watchNums[fname] == 0 {
+		err = shared.watcher.Add(fname)
 	}
-
-	err := shared.watcher.Add(fname)
 	if err == nil {
 		shared.watchNums[fname]++
-		if winfo.isCreate() {
-			shared.watchNums[winfo.fname]++
-		}
 	}
 	return err
 }
 
 // removeWatch calls fsnotify.RemoveWatch for the input filename and closes the
 // corresponding events channel.
-func (shared *InotifyTracker) removeWatch(winfo *watchInfo) {
+func (shared *InotifyTracker) removeWatch(winfo *watchInfo) error {
 	shared.mux.Lock()
-	defer shared.mux.Unlock()
 
 	ch := shared.chans[winfo.fname]
-	if ch == nil {
-		return
+	if ch != nil {
+		delete(shared.chans, winfo.fname)
+		close(ch)
 	}
 
-	delete(shared.chans, winfo.fname)
-	close(ch)
-
-	if !winfo.isCreate() {
-		return
+	fname := winfo.fname
+	if winfo.isCreate() {
+		// Watch for new files to be created in the parent directory.
+		fname = filepath.Dir(fname)
+	}
+	shared.watchNums[fname]--
+	watchNum := shared.watchNums[fname]
+	if watchNum == 0 {
+		delete(shared.watchNums, fname)
 	}
+	shared.mux.Unlock()
 
-	shared.watchNums[winfo.fname]--
-	if shared.watchNums[winfo.fname] == 0 {
-		delete(shared.watchNums, winfo.fname)
+	var err error
+	// If we were the last ones to watch this file, unsubscribe from inotify.
+	// This needs to happen after releasing the lock because fsnotify waits
+	// synchronously for the kernel to acknowledge the removal of the watch
+	// for this file, which causes us to deadlock if we still held the lock.
+	if watchNum == 0 {
```

Comment:
All `addWatch` and`removeWatch` calls are made from a single [goroutine](https://github.com/vsco/tail/pull/2/files/2e79377851962f5983a4cd8ccda689920fb6ef9f#diff-55c813f720796366a2276f73f7ade8a4L238), so access is serialized. As long as the calls are properly ordered (`add` < corresponding `remove`), the result should be correct.